src/librustc_resolve/build_reduced_graph.rs

```diff
@@ -337,8 +337,20 @@ impl<'a> Resolver<'a> {
 
             // These items live in both the type and value namespaces.
             ItemKind::Struct(ref struct_def, _) => {
+                if self.session.features.borrow().non_exhaustive {
+                    let has_non_exhaustive = item.attrs.iter()
+                        .any(|item| item.check_name("non_exhaustive"));
+
+                    // If the structure is marked as non_exhaustive then lower the visibility
+                    // to within the crate.
+                    if has_non_exhaustive && vis == ty::Visibility::Public {
+                        vis = ty::Visibility::Restricted(DefId::local(CRATE_DEF_INDEX));
```

Comment:
Hmm, this doesn't look correct - constructor's visibility below (`ctor_vis`) should be lowered to `pub(crate)`, not visibility of the struct itself.

Reply:
Adjusted this.